### PR TITLE
KIALI-563: new real browser smoke test (part 1 & 2)

### PR DESCRIPTION
**Pull request by vnugent**
Part 1 (this PR)
- [x] access full app in headless Chrome
- [x] new cmd  `npm run kiali` and `npm run itest` 
- [x] take a screenshot. upload image to  https://github.com/kiali-bot/travis-screenshots
- [x] post a comment to the PR with embedded screenshot

Part 2 (~~future~~ this PR)
- [x] created a simple proxy app directing API calls to the public Kiali instance

To test this feature locally:
```
# npm install
# npm start
# npm run itest
```
a screenshot file `kiali-browser-test.png` will be generated in the project base dir.

new npm commands:
- `npm run kiali`: launch Kiali UI serving static files from ./build dir; API calls are proxy'ed to http://kiali-istio-system.45.56.100.74.nip.io
- `npm run itest`: run only integration tests in `__itests__` dir

Note: Since travis-ci doesn't expose encrypted env in PR build for security reason, I had to resort to a workaround, putting a github token in travis.yml.  It's a separate account with id 'kiali-bot'.  If you see  strange activities by kiali-bot, please let me know.

image

kiali-browser-test

### CODE REVIEW

**Comment by josejulio** [comment]
Hola!

File: .travis.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ script:
     npm test -- --coverage
   - serve -s build &
   - npm run itest
-  - 'curl -H "Authorization: token fbc63650102b81eefdad27f363516cb105ea50b8" -X PUT -d '{"message": "Test upload1", "content": "'"$(base64 -w 0 kiali-browser-test.png)"'"}' "https://api.github.com/repos/kiali-bot/travis-screenshots/contents/kiali-ui/travis-build-${TRAVIS_BUILD_NUMBER}.png"'
+  - 'curl -s -H "Authorization: token fbc63650102b81eefdad27f363516cb105ea50b8" -X PUT -d {\"message\": \"Test upload1\", \"content\": \"$(base64 -w 0 kiali-browser-test.png)\"} "https://api.github.com/repos/kiali-bot/travis-screenshots/contents/kiali-ui/travis-build-${TRAVIS_BUILD_NUMBER}.png"'
```
Comment (josejulio) [inline .travis.yml]: I think this is OK for testing, but once all is set, please delete the token, create a new one and encrypt it: https://docs.travis-ci.com/user/encryption-keys/

Comment (vnugent) [reply .travis.yml]: @josejulio 
Unfortunately it's not possible to access encrypted envs in PR as noted in the link above.  There's a long discussion on travis-ci repo about this issue, and one reasonable workaround is to create a separate read-only Github account for this purpose and obfuscate the password in .travis.yml with base64.

> Please note that encrypted environment variables are not available for pull requests from forks.

Comment (josejulio) [reply .travis.yml]: darn! right, i had the same issue and had to offload some of the work on a heroku server and be notified as a webhook.
Basically the server did read the travis log to comment a success/fail status (the password being in the server)

https://github.com/hawkular/hawkular-services/blob/master/.travis.yml#L22

File: .travis.yml [hunk 2]
```diff
@@ -29,4 +40,6 @@ after_deploy:
...
+    - secure: aI1uMj0h6xjzp2Xl8VeG4Eibnmz77OaTs8vYyOwgf4L5eN59Rz21oJE4j4AFkDBELUJvYHJix/EKWoIGszDFN/XCoOcB39h9MFyrHT74dwmjH1CHDg97e8KZU8oKpvpJ6D5c1tYUrbuzeK+pxqGptMi3zi9Cx8UpN9CavVa1QkcfEf0VfVdGpHIf+EbP9imEnArfPkUMfx0Schp761gySOvgtzzjIUwFLrrszSWM5s5KA8Z2BWMKN4877uey2HCYqfju9TrjWv396QxWxdnzqGvTQCO5jabcxSuK3N5IC76mlkuT/w0rxHVwULDU9JyBQo6gYGkHavKq5vvTOrWzEZwhL5mY6RvCwPg4GJma+O6DWZQRLEbF18Ubu2pm8BbLp+EUdgQMN+rBhTcbhX/om+tZkyODoVmxSEibSo6LcDJ3ZiS8pzlFVQiUbQ3wiQqcIPoDwlhhahwUutd35zt26HJvSs4HuRZ1R6TU3xtsHFVQULbIOMWcWd5dFnzTdBJW3aNu45vkWOE7wtANiAKmpkFowq0lRoKj6GdHgDEk9L0xy/QmoiCmCnrqtSsIill0GjmPpXWzIBQmS4uPUkjGveIqIGfrRqnSyRa6+f0wY4GQ0vkbF4DFbCTvRVcfvvph2R3MS/Tjhvt5L5Bhw52lg/DiWib2Synle1WyIjhXTLk=
+    - CI=true
```
Comment (josejulio) [inline .travis.yml]: I think this is the default value
 https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables

Comment (vnugent) [reply .travis.yml]: Thanks. I'll remove CI=true

**Comment by vnugent** [comment]
The last screenshot https://github.com/kiali/kiali-ui/pull/316#issuecomment-387493972 is posted automatically by the integration test script

File: package.json [hunk 3]
```diff
@@ -63,12 +67,16 @@
...
+    "express": "^4.16.3",
+    "http-proxy-middleware": "^0.18.0",
     "husky": "0.14.3",
     "jest-canvas-mock": "1.0.2",
     "jest-localstorage-mock": "2.2.0",
+    "jest-puppeteer": "^3.0.1",
     "node-sass-chokidar": "1.2.2",
     "npm-snapshot": "1.0.3",
     "prettier": "1.12.1",
+    "puppeteer": "^1.3.0",
```
Comment (josejulio) [inline package.json]: Can you lock all your dependencies?

Comment (vnugent) [reply package.json]: @josejulio  removed all ~

File: run_itest.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+if [ -z ${TRAVIS_PULL_REQUEST} ] || [ ${TRAVIS_PULL_REQUEST} = "false" ]; 
+then
+  echo "Missing env TRAVIS_PULL_REQUEST.  Must be run from a PR."
```
Comment (josejulio) [inline run_itest.sh]: Wouldn't be worth to run these itest also in master branch? or these tests are going to be exclusive for posting the screenshot?

Comment (vnugent) [reply run_itest.sh]: we can run on PR build for now.  I just pushed a new small update to the if statement

File: package.json [hunk 5]
```diff
@@ -63,12 +67,16 @@
...
+    "express": "4.16.3",
+    "http-proxy-middleware": "0.18.0",
     "husky": "0.14.3",
     "jest-canvas-mock": "1.0.2",
     "jest-localstorage-mock": "2.2.0",
+    "jest-puppeteer": "3.0.1",
     "node-sass-chokidar": "1.2.2",
     "npm-snapshot": "1.0.3",
     "prettier": "1.12.1",
+    "puppeteer": "1.3.0",
```
Comment (mtho11) [inline package.json]: Ah puppeteer, I was going to ask if you guys had used it.

**Pull request merged**